cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/RemoveDisksRunnable.java

```diff
@@ -0,0 +1,159 @@
+    }
+
+    @Override
+    protected void init() {
+        _kafkaCruiseControl.sanityCheckDryRun(_dryRun, _stopOngoingExecution);
+        _goalsByPriority = new ArrayList<>(1);
+        _goalsByPriority.add(new DiskRemovalGoal(_brokerIdAndLogdirs, _errorMargin));
+
+        _operationProgress = _future.operationProgress();
+        if (_stopOngoingExecution) {
+            maybeStopOngoingExecutionToModifyAndWait(_kafkaCruiseControl, _operationProgress);
+        }
+        _combinedCompletenessRequirements = _goalsByPriority.get(0).clusterModelCompletenessRequirements();
+    }
+
+    @Override
+    protected OptimizerResult workWithClusterModel() throws KafkaCruiseControlException, NotEnoughValidWindowsException, TimeoutException {
+        Set<Integer> brokersToCheckPresence = new HashSet<>(_brokerIdAndLogdirs.keySet());
+        _kafkaCruiseControl.sanityCheckBrokerPresence(brokersToCheckPresence);
+        ClusterModel clusterModel = _kafkaCruiseControl.clusterModel(
+                DEFAULT_START_TIME_FOR_CLUSTER_MODEL,
+                _kafkaCruiseControl.timeMs(),
+                _combinedCompletenessRequirements,
+                true,
+                _allowCapacityEstimation,
+                _operationProgress
+        );
+
+        checkCanRemoveDisks(_brokerIdAndLogdirs, clusterModel);
+
+        OptimizationOptions optimizationOptions = computeOptimizationOptions(clusterModel,
+                false,
+                _kafkaCruiseControl,
+                Collections.emptySet(),
+                _dryRun,
+                _excludeRecentlyDemotedBrokers,
+                _excludeRecentlyRemovedBrokers,
+                _excludedTopics,
+                Collections.emptySet(),
+                false,
+                _fastMode
+        );
+
+        OptimizerResult result = _kafkaCruiseControl.optimizations(clusterModel, _goalsByPriority, _operationProgress, null, optimizationOptions);
+        if (!_dryRun) {
+            _kafkaCruiseControl.executeProposals(
+                    result.goalProposals(),
+                    Collections.emptySet(),
+                    isKafkaAssignerMode(_goals),
+                    0,
+                    0,
+                    1,
+                    1,
+                    SELF_HEALING_EXECUTION_PROGRESS_CHECK_INTERVAL_MS,
+                    SELF_HEALING_REPLICA_MOVEMENT_STRATEGY,
+                    _kafkaCruiseControl.config().getLong(ExecutorConfig.DEFAULT_REPLICATION_THROTTLE_CONFIG),
+                    _isTriggeredByUserRequest,
+                    _uuid,
+                    false
+            );
+        }
+
+        return result;
+    }
+
+    private void checkCanRemoveDisks(Map<Integer, Set<String>> brokerIdAndLogdirs, ClusterModel clusterModel) {
+        for (Map.Entry<Integer, Set<String>> entry : brokerIdAndLogdirs.entrySet()) {
+            Integer brokerId = entry.getKey();
+            Set<String> logDirs = entry.getValue();
```

Comment:
Maybe change the name to something like "logDirsToRemove"

Reply:
👍